File: 0036-valid-sudoku/0036-valid-sudoku.go
func isValidSudoku(board [][]byte) bool {
    
    type Pair struct{
        row int
        column int
    }

    rowDict := make(map[int]map[byte]struct{})
    columnDict := make(map[int]map[byte]struct{})
    boxDict := make(map[Pair]map[byte]struct{})

    for rowIndex, row := range board {
        for columnIndex, num := range row {
            if num == '.' {
                continue
            }

            if _, ok := rowDict[rowIndex]; !ok {
                rowDict[rowIndex] = make(map[byte]struct{})
            }
            if _, ok := columnDict[columnIndex]; !ok {
                columnDict[columnIndex] = make(map[byte]struct{})
            }
            boxKey := Pair{rowIndex/3, columnIndex/3}
            if _, ok := boxDict[boxKey]; !ok {
                boxDict[boxKey] = make(map[byte]struct{})
            }


            if _, ok := rowDict[rowIndex][num]; ok {
                return false
            } else if _, ok := columnDict[columnIndex][num]; ok {
                return false
            } else if _, ok := boxDict[boxKey][num]; ok {
                return false
            }
            
            rowDict[rowIndex][num] = struct{}{}
            columnDict[columnIndex][num] = struct{}{}
            boxDict[boxKey][num] = struct{}{}
        }
    }
    return true
}

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row_dict = collections.defaultdict(set)
        column_dict = collections.defaultdict(set)
        box_dict = collections.defaultdict(set)

        for i in range(9):
            for j in range(9):
                current = board[i][j]
                if current==".": continue
                if current in row_dict[i] or current in column_dict[j] or current in box_dict[(i//3, j//3)]:
                    return False
                else:
                    row_dict[i].add(current)
                    column_dict[j].add(current)
                    box_dict[(i//3, j//3)].add(current)

        return True

        